packages/marshal/src/encodeToSmallcaps.js

```diff
@@ -0,0 +1,495 @@
+            return BigInt(encoding);
+          }
+          case '$': {
+            const result = decodeRemotableFromSmallcaps(
+              encoding,
+              decodeFromSmallcaps,
+            );
+            if (passStyleOf(result) !== 'remotable') {
+              assert.fail(
+                X`internal: decodeRemotableFromSmallcaps option must return a remotable: ${result}`,
+              );
+            }
+            return result;
+          }
+          case '&': {
+            const result = decodePromiseFromSmallcaps(
+              encoding,
+              decodeFromSmallcaps,
+            );
+            if (passStyleOf(result) !== 'promise') {
+              assert.fail(
+                X`internal: decodePromiseFromSmallcaps option must return a promise: ${result}`,
+              );
+            }
+            return result;
+          }
+          default: {
+            assert.fail(
+              X`Special char ${q(c)} reserved for future use: ${encoding}`,
+            );
+          }
+        }
+      }
+      case 'object': {
+        if (encoding === null) {
+          return encoding;
+        }
+
+        if (isArray(encoding)) {
+          const result = [];
+          const { length } = encoding;
+          for (let i = 0; i < length; i += 1) {
+            result[i] = decodeFromSmallcaps(encoding[i]);
+          }
+          return result;
+        }
+
+        if (hasOwnPropertyOf(encoding, '#tag')) {
+          const { '#tag': tag, payload, ...rest } = encoding;
+          typeof tag === 'string' ||
+            assert.fail(
+              X`Value of "#tag", the tag, must be a string: ${encoding}`,
+            );
+          ownKeys(rest).length === 0 ||
+            assert.fail(
+              X`#tag record unexpected properties: ${q(ownKeys(rest))}`,
+            );
+          return makeTagged(tag, decodeFromSmallcaps(payload));
+        }
+
+        if (hasOwnPropertyOf(encoding, '#error')) {
+          const result = decodeErrorFromSmallcaps(
+            encoding,
+            decodeFromSmallcaps,
+          );
+          passStyleOf(result) === 'error' ||
+            assert.fail(
+              X`internal: decodeErrorFromSmallcaps option must return an error: ${result}`,
+            );
+          return result;
+        }
+
+        const result = {};
+        for (const encodedName of ownKeys(encoding)) {
+          // TypeScript confused about `||` control flow so use `if` instead
+          // https://github.com/microsoft/TypeScript/issues/50739
+          if (typeof encodedName !== 'string') {
+            assert.fail(
+              X`Property name ${q(
+                encodedName,
+              )} of ${encoding} must be a string`,
+            );
+          }
+          !encodedName.startsWith('#') ||
+            assert.fail(
+              X`Unrecognized record type ${q(encodedName)}: ${encoding}`,
+            );
+          const name = decodeFromSmallcaps(encodedName);
```

Comment:
```suggestion
          const name = decodeFromSmallcaps(encodedName);
          typeof name === 'string' ||
            assert.fail(
              X`Decoded property name ${name} from ${encoding} must be a string`,
            );
```
This is probably worth a test case as well.

Reply:
Done